golang/json/context.go

```diff
@@ -0,0 +1,73 @@
+package json
+
+// Copyright (c) 2015 Uber Technologies, Inc.
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import "golang.org/x/net/context"
+
+// Context is a JSON Context which contains request and response headers.
+type Context interface {
+	context.Context
+
+	// Headers returns the call request headers.
+	Headers() interface{}
+
+	// ResponseHeaders returns the call response headers.
+	ResponseHeaders() interface{}
+
+	// SetResponseHeaders sets the given response headers on the context.
+	SetResponseHeaders(headers interface{})
```

Comment:
Yeah it's a bit of a pain to break out as well, since if the requester passes a Context which doesn't contain a SetResponseHeaders, then the Call will have to cast it to a ServerContext so it can actually set the response headers it receives. I was hoping to use standard contexts and use WithValue for headers, but unfortunately response headers can't be set that way, since we can't add values to an existing context passed by the user.